import/index_java/src/org/vufind/index/CallNumberTools.java

```diff
@@ -70,6 +73,68 @@ public String getCallNumberLabel(final Record record, String fieldSpec) {
         }
     }
 
+    /**
+     * Extract the call number labels from a record
+     * @param record MARC record
+     * @return Call number labels
+     */
+    public List<String> getCallNumberLabels(final Record record) {
+        return getCallNumberLabels(record, "050a:082a:083a:090a");
+    }
+
+    /**
+     * Extract the call number labels from a record.
+     * fieldSpec might include 952e for FOLIO, in which case we do not know the type of call number
+     * (also sometimes call numbers are in the wrong MARC field for their type).
+     * If LCC: classLetters, classLetters + classDigits, classLetters + classDigits + classDecimal
+     * If Dewey: classDigits, classDigits + classDecimal
+     * Otherwise: everything up until the dot (or everything if there is no dot)
+     * @param record MARC record
+     * @param fieldSpec taglist for call number fields
+     * @return Call number labels
+     */
+    public List<String> getCallNumberLabels(final Record record, String fieldSpec) {
+        Set<String> values = SolrIndexer.instance().getFieldList(record, fieldSpec);
+        HashSet<String> result = new LinkedHashSet<String>();
+        for (String cn: values) {
+            String cnUp = cn.toUpperCase().trim();
+            LCCallNumber lcc = new LCCallNumber(cnUp);
+            DeweyCallNumber dewey = new DeweyCallNumber(cnUp);
+            boolean other = cnUp.contains(":") || (cnUp.length() > 10 && !cnUp.contains("."));
+            boolean isLcc = lcc.isValid() && !other;
+            boolean isDewey = dewey.isValid() && !other;
+            if (isLcc) {
+                result.add(lcc.getClassLetters());
+                result.add(lcc.getClassLetters() + lcc.getClassDigits());
+                if (lcc.getClassDecimal() != null) {
+                    result.add(lcc.getClassLetters() + lcc.getClassDigits() + lcc.getClassDecimal());
+                }
+            } else if (isDewey) {
+                result.add(dewey.getClassDigits());
+                String classDecimal = dewey.getClassDecimal();
+                if (classDecimal != null) {
+                    for (int i=1; i<5; i++) {
+                        if (classDecimal.length() > i) {
+                            result.add(dewey.getClassDigits() + classDecimal.substring(0, i+1));
+                        }
+                    }
+                }
+            } else {
+                // NOTE: we could add other classifications like SuDoc here (and add related MARC fields to fieldSpec)
+                int dotPos = cnUp.indexOf(".");
+                if (dotPos < 0) {
+                    result.add(cnUp.trim());
```

Comment:
cnUp was already trimmed above; no need to do it again.

```suggestion
                    result.add(cnUp);
```